### PR TITLE
Query greenwave with JS.

### PR DESCRIPTION
**Pull request by ralphbean**
This changes the frontend only for #2370.

Signed-off-by: Ralph Bean <rbean@redhat.com>

### CODE REVIEW

**Comment by ralphbean** [comment]
Doh, when I try to `btest` I get:

```
Running Sphinx v1.6.6
making output directory...

Extension error:
Could not import extension cornice_sphinx (exception: No module named cornice_sphinx)
make: *** [Makefile:50: html] Error 1
make: Leaving directory '/home/vagrant/bodhi/docs'
Shared connection to 192.168.121.101 closed.
```

**Comment by ralphbean** [comment]
Huh.. `flake8-2` in Jenkins flipped out with:

```
rawhide	pkg_resources.DistributionNotFound: The 'pyflakes<1.7.0,>=1.5.0' distribution was not found and is required by flake8
```

**Comment by pypingou** [comment]
In https://github.com/fedora-infra/bodhi/pull/2392 I'm adding a new API endpoint to query greenwave so that the CLI can also get this information without having to specify which greenwave instance/endpoint to query and without duplicating the logic in the server and the CLI on how to create the subjects and all.

Should we take advantage of this new endpoint here as well?

**Comment by ralphbean** [comment]
> Should we take advantage of this new endpoint here as well?

Hm.  That would reduce the number of places we have to specify that greenwave URL at the cost of increased network latency.

I leave it to you guys to decide.  I can go either way.  :)

**Comment by ralphbean** [comment]
Here's a video showing it in action:

http://threebean.org/img/live-query.webm

**Comment by bowlofeggs** [comment]
jenkies test

File: bodhi/server/templates/update.html [hunk 1]
```diff
@@ -37,62 +35,62 @@
       ABSENT: 'question-circle',
     }
 
-    // XXX - For development, we typically have updates in our development
-    // bodhi DB from a snapshot from over a year ago -- before the deployment
-    // of taskotron.  As such, taskotron doesn't know about the updates that
-    // we ask about.  Therefore, it is useful sometimes to mess around here
-    // and hardcode something like this:
-    //var update = 'ugene-1.14.2-1.fc21';
```
Comment (bowlofeggs) [inline bodhi/server/templates/update.html:47]: Getting out the old dusty broom - +1

File: bodhi/server/templates/update.html [hunk 2]
```diff
@@ -365,26 +334,17 @@
...
     $("#tab-automatedtests").append(" <span class='fa fa-spinner fa-spin'></span>");
 
-    for (index = 0, len = builds.length; index < len; ++index) {
-      request_results(base_url+"results/latest?original_spec_nvr="
-        +builds[index]+"&testcases=org.centos.prod.ci.pipeline.complete");
-      request_results(base_url+"results/latest?item="+builds[index]+"&type=koji_build&testcases:like=dist.*");
-    }
+    get_greenwave_information({
+        'product_version': '${update.product_version}',
+        'decision_context': 'bodhi_update_push_stable',
```
Comment (bowlofeggs) [inline bodhi/server/templates/update.html]: In the Bodhi backend code, there is also a query to Greenwave to assert that it's OK to push an update to testing when the update is pending. Should we do that here as well?

**Review by bowlofeggs — COMMENTED** [review]
I'm fine either way with having this patch use the new API from @pypingou's patch, or having it query Greenwave directly like this. If we go with a JS implementation, I may later refactor it to use the API, but I'm kind of on the fence. The pros and cons don't strongly sway me one way or the other.

File: bodhi/server/templates/update.html (hunk 2 above)
Comment (ralphbean) [reply bodhi/server/templates/update.html]: Hm, yes.

Comment (ralphbean) [reply bodhi/server/templates/update.html]: See f2a4919b82c685ce9038e1bd536bec214b928e32.

**Comment by bowlofeggs** [comment]
The test failures seem to be mostly (if not all) because there were tests for the removed Python function: https://ci.centos.org/job/bodhi-bodhi/1539/ so we can probably just drop those tests.

I tested the PR against production Greenwave and it seems to work, though of course production Greenwave is currently configured to pass all updates. Have you been able to test this against updates that fail Greenwave gating too?

**Comment by bowlofeggs** [comment]
Let's add a note to ```docs/user/release_notes.rst``` that mentions that Greenwave is now a required service for Bodhi deployments that wish to continue displaying test results in the UI.

**Comment by bowlofeggs** [comment]
There seems to be one failing test that is asserting that the template tells the developer that the update cannot be pushed.

**Comment by ralphbean** [comment]
> There seems to be one failing test that is asserting that the template tells the developer that the update cannot be pushed.

Ack - haven't had a chance to resolve that one yet.

**Comment by ralphbean** [comment]
> There seems to be one failing test that is asserting that the template tells the developer that the update cannot be pushed.

Should be happy now.  :)

**Comment by ralphbean** [comment]
> Have you been able to test this against updates that fail Greenwave gating too?

Yes.  :)

> I tested the PR against production Greenwave and it seems to work, though of course production Greenwave is currently configured to pass all updates.

Try staging.  There are actual requirements in the staging greenwave policies.

**Comment by ralphbean** [comment]
Rebased against `develop`.

**Comment by ralphbean** [comment]
Rebased against develop again.  :)

**Comment by bowlofeggs** [comment]
Would you care to update your commit message to include this?

```
fixes #2425
```

Seems like you saved us from another issue!

**Comment by ralphbean** [comment]
@bowlofeggs, done in e7cb406.

**Comment by ralphbean** [comment]
@bowlofeggs, @pypingou - anything else to change here?

**Review by bowlofeggs — APPROVED** [review]
I spot checked this PR against a handful of updates and I didn't notice any problems. However, I didn't find any updates that had missing or wavied tests in staging Greenwave. I will take your word that you tested those scenarios and merge, since Bodhi doesn't currently handle those scenarios well today anyway.

**Comment by bowlofeggs** [comment]
Thanks for the patch!

**Comment by bowlofeggs** [comment]
These patches are planned to be included in the upcoming 3.9.0 release: #2477.

**Comment by bowlofeggs** [comment]
These patches were deployed to Fedora's production Bodhi instance yesterday.

**Comment by AdamWill** [comment]
I think this broke display of openQA results. At least, it's the best suspect I have. openQA is still happily testing updates and forwarding the results to resultsdb, but Bodhi's Automated Tests tab is not showing them. e.g. [here](https://bodhi.fedoraproject.org/updates/FEDORA-2018-87ba0312c2) - there certainly are [results for that update in ResultsDB](https://taskotron.fedoraproject.org/resultsdb/results?item:like=FEDORA-2018-87ba0312c2*).

At a quick glance it looks like this changed Bodhi from querying resultsdb directly to querying Greenwave and relying on Greenwave to give it all the relevant results from resultsdb. I guess Greenwave is not picking up openQA results at present. :/

**Comment by bowlofeggs** [comment]
Yeah looks like Greenwave isn't returning that data. No match here:

```
curl 'https://greenwave-web-greenwave.app.os.fedoraproject.org/api/v1.0/decision' -H 'User-Agent: Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:63.0) Gecko/20100101 Firefox/63.0' -H 'Accept: */*' -H 'Accept-Language: en-US,en;q=0.5' --compressed -H 'Content-Type: application/json' -H 'Origin: https://bodhi.fedoraproject.org' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Cache-Control: max-age=0' --data '{"product_version":"fedora-29","decision_context":"bodhi_update_push_stable","subject":[{"item":"kernel-4.19.5-300.fc29","type":"koji_build"},{"item":"kernel-headers-4.19.5-300.fc29","type":"koji_build"},{"item":"kernel-tools-4.19.5-300.fc29","type":"koji_build"},{"original_spec_nvr":"kernel-4.19.5-300.fc29"},{"original_spec_nvr":"kernel-headers-4.19.5-300.fc29"},{"original_spec_nvr":"kernel-tools-4.19.5-300.fc29"},{"item":"FEDORA-2018-87ba0312c2","type":"bodhi_update"}],"verbose":true}' | grep -i openqa
```

**Comment by AdamWill** [comment]
Yeah, I think greenwave has simply never considered openqa results because we never tried to gate on them. It was set up to pull in taskotron and pipeline results as we tried to do gating based on tests from both of those, but not openqa. So when bodhi stopped querying resultsdb directly and started just querying greenwave, we lost the openqa results.

The fix should be a change to the fedora greenwave policy in infra ansible git, I think. I'll try and figure that out this afternoon.

**Comment by AdamWill** [comment]
[Done](https://infrastructure.fedoraproject.org/cgit/ansible.git/commit/?id=0bc53e1122fdfaea3b9e670a64a01489613ccfea), and it works. yay.

**Pull request merged**